server/schedule/placement/rule_manager.go

```diff
@@ -273,3 +274,53 @@ func (m *RuleManager) FitRegion(stores StoreSet, region *core.RegionInfo) *Regio
 	rules := m.GetRulesForApplyRegion(region)
 	return FitRegion(stores, region, rules)
 }
+
+func (m *RuleManager) swapRule(rule *Rule) *Rule {
+	old := m.rules[rule.Key()]
+	m.rules[rule.Key()] = rule
+	return old
+}
+
+// SetRules inserts or updates lots of Rules at once.
+func (m *RuleManager) SetRules(rules []*Rule) error {
+	for _, rule := range rules {
+		err := m.adjustRule(rule)
+		if err != nil {
+			return err
+		}
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	oldRules := make(map[[2]string]*Rule)
+
+	for _, rule := range rules {
+		oldRules[rule.Key()] = m.swapRule(rule)
```

Comment:
May `oldRules` be nils? will it build a `rulelist` successfully if `oldRules` contains nil pointer? we need to add a test for it.

Reply:
Yes. The first case, actually, should already have `oldRules` full of nils. But its ID is duplicated with `SetRule`.